cupyx/jit/cg.py

```diff
@@ -284,5 +305,130 @@ def call_const(self, env):
         return _Data(f'cg::this_{self.group_type}()', cg_type)
 
 
+class _Sync(_BuiltinFunc):
+
+    def __call__(self):
+        """Calls ``cg::sync()``.
+
+        Args:
+            group: a valid cooperative group
+
+        .. seealso:: `cg::sync`_
+
+        .. _cg::sync:
+            https://docs.nvidia.com/cuda/archive/11.6.0/cuda-c-programming-guide/index.html#collectives-cg-sync
+        """
+        super().__call__()
+
+    def call(self, env, group):
+        _check_cg_include(env)
+        return _Data(f'cg::sync({group.code})', _cuda_types.void)
+
+
+class _MemcpySync(_BuiltinFunc):
+
+    def __call__(self):
+        """Calls ``cg::memcpy_sync()``.
+
+        Args:
+            group: a valid cooperative group
+            dst: the destination array that can be viewed as a 1D
+                C-contiguous array
+            dst_idx: the start index of the destination array element
+            src: the source array that can be viewed as a 1D C-contiguous
+                array
+            src_idx: the start index of the source array element
+            size (int): the number of bytes to be copied from
+                ``src[src_index]`` to ``dst[dst_idx]``
+            aligned_size (int): Use ``cuda::aligned_size_t<N>`` to guarantee
+                the compiler that ``src``/``dst`` are at least N-bytes aligned.
+                The behavior is undefined if the guarantee is not held.
+
+        .. seealso:: `cg::memcpy_sync`_
+
+        .. _cg::memcpy_sync:
+            https://docs.nvidia.com/cuda/archive/11.6.0/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async
+        """
+        super().__call__()
+
+    def call(self, env, group, dst, dst_idx, src, src_idx, size, *,
```

Comment:
Note: this signature is borrowing from the atomic functions. It seems each pointer address argument in C maps to two Python arguments (`array` & `index`). 

cc: @gmarkall in case Numba is planning on expanding the support for cooperative groups 🙂 (CuPy borrowed the atomic function signature from Numba, and it'd be nice to keep the signature consistent.)

Reply:
Thanks for the CC - my plan for extending Numba support for this is to just enable directly calling the C++ equivalents from Numba kernels once this is merged: https://github.com/numba/numba/pull/7621

Reply:
I see, so the planned approach is to ask Numba users to write CG-related codes directly in C++, and then link to a Numba kernel?

Reply:
Ah, sorry - I didn't quite represent my intention properly - I'd write some C shim functions that call the C++ versions (similar to https://github.com/numba/numba-examples/pull/40/files#diff-e494235cb05f7f751143c66858983a68f1b7630a7a186154de92f213d80179a0) then have that shim automatically linked in by Numba if the user calls them. They'd write their CG-related code in Numba, but internally it would be using the C++ functions underneath it. 

We'd have to map pointer address arguments to the base and offset in Python for these as well (similar to what I'm doing for `curand_init` here: https://github.com/numba/numba-examples/pull/40/files#diff-e494235cb05f7f751143c66858983a68f1b7630a7a186154de92f213d80179a0R23, called like https://github.com/numba/numba-examples/pull/40/files#diff-14c8f7166d95bffa81bda8cdc5b386c5d47737132f63794019e386814464d4d3R32 from Python.